libs/rpc/connection.js

```diff
@@ -43,18 +69,86 @@ Connection.prototype.notify = function (method) {
  * @param  {Function} callback
  */
 Connection.prototype.on = function (method, callback) {
-	this._listeners[method] = callback;
+	if (this._callbacks[method]) {
+		throw new Error("Method already registered: " + method);
+	}
+
+	this._callbacks[method] = callback;
+};
+
+Connection.prototype._send = function (obj) {
+	if (obj) {
+		var msg = this._protocol.serialize(obj);
+		this._driver.send(msg);
+	}
 };
 
 /**
- * Handler for request sent from remote app
- * @param  {Object} request
- * @return {Object}
+ * Handler for messages sent from remote app
+ * @param  {Object} msg
  */
-Connection.prototype._onRecieve = function (request) {
-	var callback = this._listeners[request.method];
+Connection.prototype._onRecieve = function (msg) {
+	var obj = this._protocol.parse(msg);
+
+	if (obj instanceof Request) {
+		this._handleRequest(obj);
+	}
+	else if (obj instanceof Response) {
+		this._handleResponse(obj);
+	}
+};
+
+/**
+ * Handle request from remote app
+ * @param  {Resuest} req
+ */
+Connection.prototype._handleRequest = function (req) {
+	var response = null;
+
+	var callback = this._callbacks[req.method];
 	if (callback) {
-		return callback(request.data);
+		// Handle notification
+		if (req.isNotification) {
+			callback.apply(null, req.args);
+		}
+		// Handle request
+		else {
+			// Call method and return response
+			try {
+				var result = callback.apply(null, req.args);
+				response = Response.createResponse(req.id, result);
+			}
+			// Return back error
+			catch (err) {
+				response = Response.createError(req.id, err.toString());
+			}
+		}
+	}
+	// Called method does not exist
+	else {
+		response = Response.createError(req.id, "Method does not exist: " + req.method);
+	}
+
+	// Send response if exists
+	this._send(response);
+};
+
+/**
+ * Handle response from remote app
+ * @param  {Response} resp
+ */
+Connection.prototype._handleResponse = function (resp) {
+	var promise = this._promises[resp.id];
+
+	if (promise) {
```

Comment:
I would avoid this if, since it will allow silent passing of errors. If we get response, but we do not have promise which is waiting for it, we cannot detect that since there is no else block. I would just remove this if and exception will be throw if  undefined.reject is called.